### PR TITLE
entityUpdaters

### PR DESCRIPTION
**Pull request by nevir**
update helpers on a per-typename basis, declared globally.  This allows you to guarantee that derived fields are kept in sync, even across queries.

### CODE REVIEW

File: src/CacheTransaction.ts [hunk 1]
```diff
@@ -95,6 +104,41 @@ export class CacheTransaction implements Queryable {
...
+    const nodesToEmit = [];
+    for (const nodeId of this._editedNodeIds) {
+      const node = nextSnapshot.getNodeSnapshot(nodeId);
+      const previous = prevSnapshot.getNodeSnapshot(nodeId);
+      // One of them may be undefined; but we are guaranteed that both represent
+      // the same entity.
+      const either = node || previous;
+
+      if (!(either instanceof EntitySnapshot)) continue; // Only entities
+      const typeName = isObject(either.data) && either.data.__typename as string | undefined;
```
Comment (yuit) [inline src/CacheTransaction.ts]: Should this be typename or "Query" for the rootQuery ?

File: src/CacheTransaction.ts [hunk 2]
```diff
@@ -95,6 +104,41 @@ export class CacheTransaction implements Queryable {
...
+      if (!updater) continue;
+
+      nodesToEmit.push([updater, node && node.data, prevSnapshot.getNodeData(nodeId)]);
+    }
+
+    if (!nodesToEmit.length) return;
+
+    // TODO: This is weirdly the only place where we assume an Apollo interface.
+    // Can we clean this up? :(
+    const dataProxy = new ApolloTransaction(this);
```
Comment (yuit) [inline src/CacheTransaction.ts]: Why do we have to create ApolloTrasaction here?  I thought we are already in one

File: src/CacheTransaction.ts (hunk 1 above)
Comment (nevir) [reply src/CacheTransaction.ts]: Doh, yeah - should be `Query` for the root node

File: src/CacheTransaction.ts (hunk 2 above)
Comment (nevir) [reply src/CacheTransaction.ts]: Nope, we're in a `CacheTransaction` :(

Comment (nevir) [reply src/CacheTransaction.ts]: But in order to keep the interface consistent, gotta expose it as an apollo transaction to the callbacks.

This needs some fixing, but I think that's going to require a fair bit of refactoring to the public interfaces (as well as Apollo's own interfaces)

**Pull request merged**